File: plugins/mobile-interaction/ios/Sources/MobileInteractionPlugin.swift
import Foundation
import Tauri
import UIKit

private class HapticArgs: Decodable {
  let style: String
}

class MobileInteractionPlugin: Plugin {
  @objc public func perform(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(HapticArgs.self)

    DispatchQueue.main.async {
      switch args.style {
      case "selection":
        let generator = UISelectionFeedbackGenerator()
        generator.prepare()
        generator.selectionChanged()
      case "impact":
        let generator = UIImpactFeedbackGenerator(style: .medium)
        generator.prepare()
        generator.impactOccurred()
      case "success":
        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(.success)
      case "warning":
        let generator = UINotificationFeedbackGenerator()
        generator.prepare()
        generator.notificationOccurred(.warning)
      default:
        invoke.reject("Unsupported haptic style: \(args.style)")
        return
      }
      NSLog("JTypeHaptics style=%@ performed=true", args.style)
      invoke.resolve(["performed": true])
    }
  }
}

@_cdecl("init_plugin_mobile_interaction")
func initPlugin() -> Plugin {
  MobileInteractionPlugin()
}

File: plugins/mobile-interaction/android/src/main/java/net/jcode/jtype/mobileinteraction/MobileInteractionPlugin.kt
package net.jcode.jtype.mobileinteraction

import android.app.Activity
import android.os.Build
import android.util.Log
import android.view.HapticFeedbackConstants
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
class HapticArgs {
    lateinit var style: String
}

@TauriPlugin
class MobileInteractionPlugin(private val activity: Activity) : Plugin(activity) {
    @Command
    fun perform(invoke: Invoke) {
        val args = try {
            invoke.parseArgs(HapticArgs::class.java)
        } catch (error: Exception) {
            invoke.reject("Invalid haptic request: ${error.message}")
            return
        }

        activity.runOnUiThread {
            val feedback = when (args.style) {
                "selection" -> HapticFeedbackConstants.CLOCK_TICK
                "impact" -> HapticFeedbackConstants.CONTEXT_CLICK
                "success" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    HapticFeedbackConstants.CONFIRM
                } else {
                    HapticFeedbackConstants.KEYBOARD_TAP
                }
                "warning" -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    HapticFeedbackConstants.REJECT
                } else {
                    HapticFeedbackConstants.LONG_PRESS
                }
                else -> {
                    invoke.reject("Unsupported haptic style: ${args.style}")
                    return@runOnUiThread
                }
            }
            val performed = activity.window.decorView.performHapticFeedback(feedback)
            Log.d("JTypeHaptics", "style=${args.style} performed=$performed")
            invoke.resolve(JSObject().put("performed", performed))
        }
    }
}
